programs/host/frontend.c

```diff
@@ -0,0 +1,957 @@
+                uint32_t timeout = item_uint32( inputs[2] );
+                ALLOC_OUTPUT( 0 , len ); // data
+                if( timeout == MBEDTLS_SERIALIZE_TIMEOUT_INFINITE )
+                {
+                    DBG( "executing recv fd=%u len=%u", (int) ctx.fd, (unsigned) len );
+                    ret = mbedtls_net_recv( &ctx, item_buffer( outputs[0] ), len );
+                }
+                else
+                {
+                    DBG( "executing recv_timeout fd=%u len=%u timeout=%u",
+                        (unsigned) ctx.fd, (unsigned) len, (unsigned) timeout );
+                    ret = mbedtls_net_recv_timeout( &ctx,
+                            item_buffer( outputs[0] ), len,
+                            timeout );
+                }
+                if( ret >= 0 )
+                {
+                    DBG( "received %zu bytes on fd=%d", (size_t) ret, (int) ctx.fd );
+                    outputs[0]->size = ret;
+                    ret = 0;
+                }
+            }
+            break;
+
+        case MBEDTLS_SERIALIZE_FUNCTION_SEND:
+            CHECK_ARITY( 2 );
+            CHECK_LENGTH( 0, 2 ); // fd
+            {
+                mbedtls_net_context ctx = { item_uint16( inputs[0] ) };
+                size_t len = inputs[1]->size;
+                unsigned char *buf = item_buffer( inputs[1] );
+                ALLOC_OUTPUT( 0 , 4 ); // sent_len
+                DBG( "executing send fd=%u len=%zu", (int) ctx.fd, len );
+                ret = mbedtls_net_send( &ctx, buf, len );
+                if( ret >= 0 )
+                {
+                    DBG( "sent %zu bytes on fd=%d", (size_t) ret, (int) ctx.fd );
+                    set_item_uint32( outputs[0], ret );
+                    ret = 0;
+                }
+            }
+            break;
+
+        case MBEDTLS_SERIALIZE_FUNCTION_SHUTDOWN:
+            CHECK_ARITY( 1 );
+            CHECK_LENGTH( 0, 2 ); // fd
+            {
+                mbedtls_net_context ctx = { item_uint16( inputs[0] ) };
+                DBG( "executing shutdown fd=%d", (int) ctx.fd );
+                mbedtls_net_free( &ctx );
+                ret = 0;
+            }
+            break;
+
+        case MBEDTLS_SERIALIZE_FUNCTION_FOPEN:
+            {
+                char * path = NULL;
+                char * mode = NULL;
+                int32_t file_id;
+
+                ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+                mode = item_buffer( inputs[0] );
+                path = item_buffer( inputs[1] );
+                DBG( "open file [%s] mode [%s]", path, mode);
+                file_id = alloc_file_context();
+                DBG( "allocated file id [%d]", file_id);
+                if ( file_id != -1 )
+                {
+                    FILE * file = fopen( path, mode );
+                    if ( file != NULL )
+                    {
+                        files[file_id - 1].file = file;
+                        ALLOC_OUTPUT( 0, sizeof (int32_t) );
```

Comment:
`ALLOC_OUTPUT` can fail, so it needs to be done before `alloc_file_context` and `fopen`. In general, anything that can fail needs to be done before performing the side effects requested by the function. If you call `ALLOC_OUTPUT` and then `fopen` fails, there's no need to clean up anything, the generic clean-up code at the end of the function will take care of it (it frees all outputs if `ret != 0`).